### ISSUE TITLE
Support building and linking extra modules into sample programs and tests

### ISSUE THREAD

**Issue by gilles-peskine-arm** [ARMmbed/mbedtls#3325]
Our build system (`*/Makefile` and `*/CMakeLists.txt`) is set up to link unit tests (`tests/suites/test_suite_*`) and sample and test programs (`programs/*/*`) with the library modules (`lib/libmbed*.a`) and with any extra libraries passed in `LDFLAGS`. But there are many cases where we'd like to link extra code.

* We sometimes want to test how the library behaves with a plug-in of some type. This has come up before with `MBEDTLS_PARAM_FAILED`. It will be required to test the driver interface with statically declared drivers.
* There's some shared code for unit tests which is currently in `.function` files that are `#include`d into each C file. It would make sense for these to be separate C modules with their own `.c` and `.h`. This would speed up the compilation of tests a little and make them a bit easier to read and debug.

Goals:

* Support compiling and linking extra `.c` files residing under `tests` when building unit tests and sample programs. This code should be linked statically even if the Mbed TLS libraries are linked dynamically.
* Support both Make and CMake.
* Note that `make lib programs` should still work and should not build the unit tests, only the extra test support code if required. In the default configuration, sample programs should not require any extra code from under `tests`; this can be done either by checking the configuration from the makefiles or by guarding all the extra test code with non-default options.
* Move `mbedtls_param_failed` from `tests/suites/helpers.function` to a separate `.c` file under `tests` and link both unit tests and sample programs with that. The function must be included in the build only when `MBEDTLS_CHECK_PARAMS`, or specifically only in `component_test_check_params_functionality`. Change `component_test_check_params_functionality` to also build the sample programs (change `make lib test` to `make all test`).

Non-goals:

* Visual Studio is out of scope of this task.
* This is primarily for our own testing. It could also potentially be useful to people who port Mbed TLS to platforms that require additional platform support functions, and who use our build system. For this use case, the extra functions should be easy to link into applications that use the library. This could mean linking the extra functions into the existing `.a/`.so`, or adding a fourth one. This may require further discussion on what the public interface should be. This is beyond the scope of this task, but if it can be done easily it would be a nice bonus.

**Comment by ronald-cron-arm** [comment]
I've started to look at that. To test and demonstrate that we can compile and link .c files residing in `tests` when building unit tests, I am rather thinking about moving the whole helpers.function into a C file and include file(s) under  `tests`. The function `mbedtls_param_failed` modifies unit tests specific data thus it is not really possible to move it on its own (but if we redesign it significantly). Does this seem reasonable?

**Comment by gilles-peskine-arm** [comment]
Yes, I think changing `helpers.function` to a `.c` file is a good start.

We can give `param_fail_jmp` and `test_info` external linkage, this way there's no need to do any refactoring.

**Comment by ronald-cron-arm** [comment]
#3387 implements partially this issue then reopening it.

**Comment by ronald-cron-arm** [comment]
With #3387 (merged) and #3438 (in review) I think I've covered most of this issue. The missing part is the enablement of check parameters for programs as the implementation has gone quite far from `mbedtls_param_failed`. Do we still care about this?

**Comment by mpg** [comment]
I'm not sure I understand what the question/issue is with `mbedtls_param_failed`, could you clarify?

**Comment by ronald-cron-arm** [comment]
It is about one of the initial goals in the issue description:

- Move mbedtls_param_failed from tests/suites/helpers.function to a separate .c file under tests and link both unit tests and sample programs with that. The function must be included in the build only when MBEDTLS_CHECK_PARAMS, or specifically only in component_test_check_params_functionality. Change component_test_check_params_functionality to also build the sample programs (change make lib test to make all test).

Eventually, the sharing of code between tests and programs is not demonstrated with 'mbedtls_param_failed()' but rather with 'mbedtls_test_unhexify'. I didn't work on 'mbedtls_param_failed()' as it cannot be used in programs as it is (too unit tests machinery specific). The reworking is a bit out of the scope of this issue I'd say thus I don't plan to do it but if asked otherwise. I hope this clarifies the question.

**Comment by mpg** [comment]
Yes, thanks, this is much clearer now! I tend to agree that reworking `mbedtls_param_failed()` to be able to share it between programs and tests is a bit out of scope, and that what you did with `mbedtls_test_uhexify()` is a sufficient demonstration of code sharing between tests and programs.

Unless Gilles had something more in mind, of course.

**Comment by gilles-peskine-arm** [comment]
I do still care about `mbedtls_param_failed`. The way it's done has been a pain in the past (e.g. https://github.com/ARMmbed/mbedtls/pull/2697 needed to unblock https://github.com/ARMmbed/mbedtls/pull/1622). We used to have a definition of `mbedtls_param_failed` in every sample program, which made their maintenance harder and in particular made it hard for an external contributor to understand everything they needed to do to get a sample program to build. Now we have an `all.sh` component that avoids building sample programs because they can't be built in that configuration.

Now that all the ground work has been laid, moving `mbedtls_param_failed` to the test library, arranging it to do something sensible when called from a sample program, and changing `component_test_check_params_functionality` to build the sample programs should be easy.

**Comment by ronald-cron-arm** [comment]
ok, will do it in another PR.

**Comment by ronald-cron-arm** [comment]
#3387 and #3438 merged now but reopening as the work around `mbedtls_param_failed` is still do be done.
